File: C/Exercice5_3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

typedef struct
{
    int *tab;
    int taille;
    int x;
} TypeTableau;

void *generer(void *arg)
{
    int taille = *((int *)arg);

    TypeTableau *t = malloc(sizeof(TypeTableau));
    t->taille = taille;
    t->tab = malloc(taille * sizeof(int));

    for (int i = 0; i < taille; i++)
    {
        t->tab[i] = rand() % 100;
    }

    return (void *)t;
}

void *rechercher(void *arg)
{
    TypeTableau *t = (TypeTableau *)arg;

    int *result = malloc(sizeof(int));
    *result = 0;

    for (int i = 0; i < t->taille; i++)
    {
        if (t->tab[i] == t->x)
        {
            *result = 1;
            break;
        }
    }

    return (void *)result;
}

int main()
{
    pthread_t th1, th2;
    TypeTableau *t;
    int taille, x;
    int *res;

    srand(time(NULL));

    printf("Donner la taille du tableau : ");
    scanf("%d", &taille);

    // Thread génération
    pthread_create(&th1, NULL, generer, &taille);

    // Lecture de x en parallèle
    printf("Donner x : ");
    scanf("%d", &x);

    // Attendre la fin du thread 1
    pthread_join(th1, (void **)&t);

    // Injecter x dans la structure
    t->x = x;

    // Thread recherche
    pthread_create(&th2, NULL, rechercher, t);

    pthread_join(th2, (void **)&res);

    // Affichage
    printf("Tableau : ");
    for (int i = 0; i < t->taille; i++)
    {
        printf("%d ", t->tab[i]);
    }
    printf("\n");

    if (*res == 1)
        printf("x est dans le tableau\n");
    else
        printf("x n'est pas dans le tableau\n");

    // Libération
    free(t->tab);
    free(t);
    free(res);

    return 0;
}

File: Python/Exercice5_3.py
#coding: utf-8
import threading
import random
import time

class TypeTableau:
    def __init__(self, taille, x=None):
        self.taille = taille
        self.x = x
        self.tab = None

def generer(taille):
    t = TypeTableau(taille)
    t.tab = [random.randint(0, 99) for _ in range(taille)]
    return t

def rechercher(t):
    for val in t.tab:
        if val == t.x:
            return True
    return False

def th_generer():
    global t
    t = generer(taille)

random.seed(time.time())
taille = int(input("Donner la taille du tableau : "))
x = int(input("Donner x : "))

t = None

thread_gen = threading.Thread(target=th_generer)
thread_gen.start()
thread_gen.join()

t.x = x
res = rechercher(t)

print("Tableau :", " ".join(map(str, t.tab)))
if res:
    print("x est dans le tableau")
else:
    print("x n'est pas dans le tableau")